### ISSUE TITLE
General: Counting the IDs (an object should just have one id) leads to wrong number of objects/records

### ISSUE THREAD

**Issue by tschleider** [silknow/converter#27]
Joconde, Imatex, Garin and Ceres. 

Need some double check on the ID mechanism in the converter to avoid duplicates etc.. 
In case of Joconde we have two IDs per records because of referencing, maybe we need to re-write this.

Example query for Joconde (should lead to 366 rows):

```sparql
prefix crmdig: <http://www.ics.forth.gr/isl/CRMext/CRMdig.rdfs/>
prefix crmsci: <http://www.ics.forth.gr/isl/CRMsci/>
SELECT DISTINCT ?dig
WHERE {
 GRAPH <http://data.silknow.org/joconde> {
    ?dig a crmdig:D1_Digital_Object . }
} 
```

**Comment by tschleider** [comment]
I added you, @pasqLisena , to it, because I think we need 4 eyes on this. Please also have a look at my slides with stats about the number of records etc. as they are uploaded and retrievable from Virtuoso. 

https://docs.google.com/presentation/d/1ch8WnzBTkgDjh2ybWHEPrNV8PT3A_moMGn6FDI_vSv4/edit?usp=sharing

**Comment by pasqLisena** [comment]
It is possibly related to this: https://lists.eurecom.fr/sympa/arc/silknow/2019-01/msg00067.html

When 2 IDs (inside the same institution) are the same, the 2nd one is never converted.

What is most worrying is the behaviour of CER (2 over 3 missings).
We need examples @tschleider 

dataset|count|expected
-- | -- | --
IMATEX | 6124 | 6802
Garin | 3064 | 3109 
Joconde | 131 | 368 
CER | 410|  1291

**Comment by rtroncy** [comment]
For GARIN, there are indeed 45 items that re-use identifiers previously assigned to other objects. We cannot do anything for this.

For the other sources, this should not happen. We therefore need to investigate what is happening in the converter (or before, in the crawler)

**Comment by tschleider** [comment]
To be honest, with our worst dataset here (CER) I am at a complete loss of what to do. 

1) I just cleared the old CERES graph and re-uploaded the same dump.

2) I can show you very similar ttl files with exactly the same format, but for one (record) there is a digital Object and everything else in the KG after the upload for the other consistently nothing.

24964_MNAD.ttl and 24966_MNAD.ttl ( https://github.com/silknow/converter/blob/master/output/cer.tar.gz )

3) Try searching for 24964_MNAD and 24966 _MNAD on data.silknow.org/fct and you see what I mean.

4) If I do not query for Man-made-object nodes, but Digital Objects i get at least 200 more IDs (excluding the mentioned 24964), which still means only every second record is in the KG.

**Comment by rtroncy** [comment]
@tschleider I have tried to follow your examples, but please, include complete links, otherwise it is hard to follow.
- http://data.silknow.org/object/c938bc48-cb1c-3a86-8579-196b14434009 is the object referenced in `24966 _MNAD.ttl` ... note that there is already a bug for the image (see the local link from http://data.silknow.org/image/9bd7e666-5c54-317a-be96-c0f879286ee2 which is wrong). The data loaded in the KG seems ok with respect to what we can see in the tutle file (the 2 observations are there, etc.)
- http://data.silknow.org/object/5ef4a323-7070-3549-81bc-6ad539e78567 is the object referenced in `24964_MNAD` ... in this case, the observations, although present in the turtle file, are not loaded.

Q: did you check the log on Virtuoso? Did you try to just load this file to see whether there is not an error? Virtuoso sometimes is more strict on syntax and you may have some characters that disturb Virtuoso.

Digging a bit more in the graph, I see strange things. See http://data.silknow.org/CER and the values of the properties P106, you will see that there are big inconsistencies.

**Comment by pasqLisena** [comment]
> Q: did you check the log on Virtuoso? Did you try to just load this file to see whether there is not an error? Virtuoso sometimes is more strict on syntax and you may have some characters that disturb Virtuoso.

It's not Virtuoso that is strict, it's the TTL that is INVALID.
Virtuoso loads everything correctly until line 57, where is written:
```sparql
        ecrm:P2_has_type     "es"@Clasificación Genérica ;
```

That space breaks everything. All triples after line 57 are ignored.

**Comment by tschleider** [comment]
Thank you very much @pasqLisena , I was changing the position of the language tag anyway for #35, so this is hopefully solved after the re-conversion today!

**Comment by tschleider** [comment]
CER has now 1296 records which is exactly the same amount than the most recent crawl (1296 json files), which means it's  solved! 

For Garin and Joconde it's more or less the same.

dataset|count|expected
-- | -- | --
Garin | 3101 | 3109 
Joconde | 376 | 378 
CER | 1296|  1296

I had some delay with converting IMATEX, but it's on its way and I am optimistic.

**Comment by tschleider** [comment]
Imatex has not changed since reconversion and reupload, but I cannot find any bugs or language tag errors so it could be for another reason. Suggesting to close this issue here.

**Comment by rtroncy** [comment]
OK. Can you open a new issue for just IMATEX? Ideally, try to include in the new issue a concrete example of a record present in the JSON data crawled but either not transformed in RDF or not loaded in the endpoint.

**Comment by tschleider** [comment]
Unfortunately, my bugfix for Imatex were not applied before Reconversion and it's highly likely that it's the same issue as in the other cases (the language tag problem). Therefore I re-open it, until the reconverted version is online.

**Comment by tschleider** [comment]
Ok actually even after bugfixing and reconverting it's the same for Imatex, need to have another look after the meeting

**Comment by tschleider** [comment]
I managed to reconvert it with the non-updated code, so it needs another round of reconversion....

**Comment by tschleider** [comment]
Ok, I tried to investigate into this, but it seems that there is still a problem with the IDs, I found now proof that the IDs given by Imatex do not always make sense, so sometimes different files with different file numbers have the same ID although they describe different objects. 

Example: 

ID 83 is given both for files 5679_en.json and 5678_en.json and therefore also in the KG. 

Another problem is that some files, e.g. 4309_en.json get the URI of another file and their ID after conversion. This file has e.g. the URI: 

<http://data.silknow.org/imatex/4527_es>.

Both problems could come from the same problem: inconsistent curation of the data by IMATEX, but maybe we can work on a workaround or try to avoid inconsistencies nevertheless.

**Comment by rtroncy** [comment]
After closest inspection, there is no reason to think that IMATEX has not a proper internal ID mechanism we can rely on. The conclusion of this investigation is that we MUST re-think the way we generate UUID, avoiding to rely on filenames but rather relying on internal ID of museums. This is high priority task.

**Comment by tschleider** [comment]
Making the UUID generation depend on the internal ID did unfortunately not change anything about how many objects we find for Imatex. I will slowly change the ID mechanism from filename to internal ID for the next reconversion, but I think for Imatex the problem still lies somewhere else.

**Comment by rtroncy** [comment]
Let's wait for your re-implementation **and** documentation regarding the way the SILKNOW UUID is generated (from museum internal IDs).

Next, can you put again a concrete example illustrating the remaining problem (or its manifestation if we cannot phrase what the problem is) for IMATEX? This is unclear to me.

**Comment by tschleider** [comment]
I implemented the way the D1 URI is created for all datasets based on internal ID. If no explicit internal ID is available we still take the filename + "_local_filename_id" as ID. 

I will comment again for documenting the problem for IMATEX.

**Comment by tschleider** [comment]
I made an analysis of the crawled IMATEX json files and it makes it clear that there is a bit of a problem with the internal IDs. We would expect 1 unique ID per object multiplied by 3 for the different files per object / languages (EN,ES,CA), but 289 of the internal IDs appear between 6 and 669 times. The ID that is used by 669 different files looks like this in the KG right now, e.g.: [Link.](https://data.silknow.org/describe/?url=http%3A%2F%2Fdata.silknow.org%2Fobject%2Fb1f21606-cc30-3134-8c32-e82c1233f9f3&sid=86).

Here is the result of my concat on the JSON files as [CSV/TXT](https://github.com/silknow/converter/files/4408107/imatex-id-problem.txt).

**Comment by pasqLisena** [comment]
I think that the problem is here:
https://github.com/silknow/converter/blob/master/src/main/java/org/silknow/converter/converters/ImatexConverter.java#L72

The id is taken from `regNum`, that apparently can be null (This should be double checked).
In this case the id is not replaced and its old value is kept.

I think that in the past we were using the first part of the filename for being sure of not colliding.
I propose to restore this solution.

**Comment by tschleider** [comment]
Thanks, you are right with the analysis. Before I opened this issue it was not only the first part of the filename but the whole filename (minus the extension but including the lanuguage code), which lead to problems, as well. What about the numerical part of the file name plus the internal ID, and of course a check if an internal ID exists at all?

**Comment by pasqLisena** [comment]
> a check if an internal ID exists at all?

This is super important. Crucial to avoid "null" strings.

> What about the numerical part of the file name plus the internal ID

Why this redundancy? The numerical part of the filename should be enough

**Comment by tschleider** [comment]
> Why this redundancy? The numerical part of the filename should be enough

See comment of @rtroncy above:
https://github.com/silknow/converter/issues/27#issuecomment-575615522

To sum it up: We tried the full filename (without extension) and it lead to problems (described in this issue), I tried to replace that with the internal ID, but for Imatex this is not enough because some IDs are missing / maybe there are even more problems.

Taken everything into account, for Imatex we need:
a) A check if an internal ID exists
b) Using solely the numerical part of the filename in addition / or at least in cases without internal ID.

Thoughts @pasqLisena @rtroncy ?

**Comment by pasqLisena** [comment]
> A check if an internal ID exists

I strongly believe that this number is the numeric part of the filename

> b) Using solely the numerical part of the filename in addition / or at least in cases without internal ID.

This has a high risk of collision (e.g. filename '234' and id '1' against filename '2341' without id).



Ok, I volunteer for making a full conversion of IMATEX on my computer, for looking at the results.
Where I can find the source files?
Are those big tar files in owncloud/Data ?

**Comment by tschleider** [comment]
> I strongly believe that this number is the numeric part of the filename

Ah wait, you mean the ID of the JSON, not the ID of the object? Then i'm also sure it should be the same, I think the crawler is coded that way. 

> This has a high risk of collision (e.g. filename '234' and id '1' against filename '2341' without id).

Ok, I see. Then let's just finally use only the JSON id only and it should be fine.

>Where I can find the source files?
Are those big tar files in owncloud/Data ?

Yes it's the archives with the word "record" in the filename there: https://silknow.uv.es/owncloud/index.php/apps/files/?dir=/Data&fileid=626898

**Comment by tschleider** [comment]
I can finally confirm after the latest reconversion that we now have 6805 digital objects for Imatex instead of something like 6124. 

Before I close I will have another look at Garin and maybe apply a similar fix to get as many digital objects as there are files..

**Comment by rtroncy** [comment]
This is time to update this issue. The following SPARQL query provides the number of distinct `E22_Man-Made_Object`  and `E38_Image` for each museum ([results](https://ada-preprod.silknow.org/sparql?default-graph-uri=&query=SELECT+DISTINCT+%3Fg+count%28%3Fo%29+as+%3FnbObj+count%28%3Fi%29+as+%3FnbImg%0D%0AWHERE+%7B+%7B%0D%0A++GRAPH+%3Fg+%7B%0D%0A+++++%3Fo+a+ecrm%3AE22_Man-Made_Object+.%0D%0A++%7D+%7D%0D%0A++UNION+%7B%0D%0A++GRAPH+%3Fg+%7B%0D%0A+++++%3Fi+a+ecrm%3AE38_Image+.%0D%0A++%7D+%7D%0D%0A%7D%0D%0AGROUP+BY+%3Fg%0D%0AORDER+BY+%3Fg%0D%0A&should-sponge=&format=text%2Fhtml&timeout=0&debug=on&run=+Run+Query+)): (**running it on the preprod server**)

```
SELECT DISTINCT ?g count(?o) as ?nbObj count(?i) as ?nbImg
WHERE { {
  GRAPH ?g {
     ?o a ecrm:E22_Man-Made_Object .
  } }
  UNION {
  GRAPH ?g {
     ?i a ecrm:E38_Image .
  } }
}
GROUP BY ?g
ORDER BY ?g
```

The [article written for the Weaving Conference](https://www.overleaf.com/project/5f9c21df161437000119567e) provides the number of records and media files for each museum. The differences are highlighted in the following table:

| Museum | Number of records | Number of E22 | Number of image files | Number of E38 | status |
| ----- | :----: | :----: | :----: | :----: | ----: |
| CERES-MCU (Museos estatales del MEC) | 1296 | 1296 |  2872 | 2872 |  **ok** |
| GARIN | 3125 | 3063 | 6556 | 6224 | -62 records; -332 images | 
| IMATEX (CDMT Terrassa) | 6802 (*3) | 6805 | 9201 | 9202 | +3 records; +1 image |
| Joconde | 376 | 376 | 375 | 375 |  **ok** |
| MAD (Musée des Arts Décoratifs) | 763 | 763 | 763 | 763 | **ok** | 
| MET (Metropolitan Museum) | 8317 | 8317 | 14208 | 14211 | +3 images |
| MFA Boston (Museum of Fine Arts) | 3297 | 3297 | 3790 | 3791 | +1 image |
| Mobilier | 1295 | 1295 | 1976 | 1976 | **ok** |
| MTMAD (Musée des Tissus de Lyon) | 663 | 663 | 2958 | 2958 | **ok** |
| Paris Musées | 265 | 265 | 369 | 369 | **ok** | 
| RISD (Rhode Island School of Design) | 3338 | 3338 | 4363 | 4363 | **ok** |
| Smithsonian | 147 | 147 | 147 | 147 | **ok** |
| UNIPA | 29+146+410 | 438 | 81+146+409 | 490 | **ok** (see also #65)  | 
| VAM (Victoria and Albert Museum) | 7747 | 7747 | 22773 | 23069 | +296 images |
| Venetian | 1180 | 1180 | 1182 | 1183 | +1 image | 
| Versailles | 73 | 73 | 144 | 144 | **ok** | 

@tschleider to provide an explanation for each discrepancy, even when it is +/- 1.

**Comment by tschleider** [comment]
- GARIN: The original data is messy. I tried not only alone, but also with the help of others to match images with the excel sheets, and some just don't add up. The number of records is actually just the number of excel sheets, but some of them do not have the "proper" format. Was unfortunately a mistake to count it like that.

- IMATEX: So the original JSON files are actually not symmetrical for every language, another mistake unfortunately. There are 6804 for CA, 6799 for ES and 6803 for EN. Right now we create one record for one object and merge the data from the different languages into one. Apparently at least one record exists in EN or ES, which does not exist in CA, that's the only explanation I have. I can check which one it is.

- Mobilier: Wrong number for the record files, it's actually 1295, therefore they align.

- UNIPA: as commented on #65 everything is actually alright. 

- VAM: Here we had some trouble with the image crawling. After many fixes @ehrhart squeezed the number of missing images under 300, but it's still pretty high. It's not a conversion problem though, probably comes from the VAM website.

- General remark about records: Except for Garin and Imatex the only reason I could find for these discrepancies was an unstable ID mechanism (which might lead to D1 or E22 not getting created). I pushed a fix, but I need more testing. I will follow-up on tnhis tomorow.

**Comment by rtroncy** [comment]
Note that for VAM, the problem is **NOT** that we are missing images but we have **more** records and **more** images than expected! Thus the **+** sign. Looking forward another update after the last bug fix in generating UUID and consequent testing.

**Comment by tschleider** [comment]
For the VAM images:

[Query 1](https://data.silknow.org/sparql?default-graph-uri=&qtxt=SELECT+%3Fg+%3Fobj1+%3Fobj2+%3Fimage+WHERE+%7B%0D%0A++GRAPH+%3Fg+%7B+%3Fobj1+a+ecrm%3AE22_Man-Made_Object+%7D%0D%0A++GRAPH+%3Fg+%7B+%3Fobj2+a+ecrm%3AE22_Man-Made_Object+%7D%0D%0A++%7B%0D%0A++++SELECT+%3Fobj1+WHERE+%7B%0D%0A++++++%3Fobj1+ecrm%3AP138i_has_representation+%3Fr1+.%0D%0A++++%7D%0D%0A++++GROUP+BY+%3Fobj1%0D%0A++++HAVING%28COUNT%28%3Fr1%29+%3D+1%29%0D%0A++%7D%0D%0A++%7B%0D%0A++++SELECT+%3Fobj2+WHERE+%7B%0D%0A++++++%3Fobj2+ecrm%3AP138i_has_representation+%3Fr2+.%0D%0A++++%7D%0D%0A++++GROUP+BY+%3Fobj2%0D%0A++++HAVING%28COUNT%28%3Fr2%29+%3D+1%29%0D%0A++%7D%0D%0A++%3Fobj1+ecrm%3AP138i_has_representation%2Fschema%3AcontentUrl+%3Fimage+.%0D%0A++%3Fobj2+ecrm%3AP138i_has_representation%2Fschema%3AcontentUrl+%3Fimage+.%0D%0A++FILTER%28%3Fobj1+%21%3D+%3Fobj2%29%0D%0A%7D%0D%0A&should-sponge=&format=text%2Fhtml&timeout=0&debug=on&run=+Run+Query+): It shows objects which only have 1 representation / image and share the same image for both. (That's because of VAM, I think some of it has to do with object sets).
[Query 2](https://data.silknow.org/sparql?default-graph-uri=&qtxt=SELECT+%3Fg+%3Fobj+%3Fimgs+WHERE+{%0D%0A++GRAPH+%3Fg+{+%3Fobj+a+ecrm%3AE22_Man-Made_Object+}%0D%0A++{%0D%0A++++SELECT+%3Fobj+(GROUP_CONCAT(DISTINCT+%3Fimage%3B+separator%3D%22|%22)+AS+%3Fimgs)+WHERE+{%0D%0A++++++%3Fobj+ecrm%3AP138i_has_representation%2Fschema%3AcontentUrl+%3Fimage+.%0D%0A++++}%0D%0A++++GROUP+BY+%3Fobj%0D%0A++}%0D%0A++FILTER(!CONTAINS(%3Fimgs%2C+%22https%3A%2F%2Fsilknow.org%2Fsilknow%2Fmedia%2F%22))%0D%0A}%0D%0A&should-sponge=&format=text%2Fhtml&timeout=0&debug=on&run=+Run+Query+): A list of all the objects with an image class but no silknow link. (Here the scraping / download just failed)

If we deduct the result from the second query (59) from the one of the first query (328), we have exactly 296. So the number of images we could not download so far is relatively low, actually.

I'm still looking into the VAM records / objects themselves, there's also the discrepancy for Venetian left.

**Comment by tschleider** [comment]
Ok, good news, for both VAM and Venezia the discrepancy for the number of records / number of E22 classes seems to be solved on stage, it was probably a temporary encoding problem: 

[LINK](https://ada-preprod.silknow.org/sparql?default-graph-uri=&query=SELECT+DISTINCT+%3Fg+count(%3Fo)+as+%3FnbObj+count(%3Fi)+as+%3FnbImg%0D%0AWHERE+{+{%0D%0A++GRAPH+%3Fg+{%0D%0A+++++%3Fo+a+ecrm%3AE22_Man-Made_Object+.%0D%0A++}+}%0D%0A++UNION+{%0D%0A++GRAPH+%3Fg+{%0D%0A+++++%3Fi+a+ecrm%3AE38_Image+.%0D%0A++}+}%0D%0A}%0D%0AGROUP+BY+%3Fg%0D%0AORDER+BY+%3Fg%0D%0A&should-sponge=&format=text%2Fhtml&timeout=0&debug=on&run=+Run+Query+)

**Comment by rtroncy** [comment]
Good, I have update the table in https://github.com/silknow/converter/issues/27#issuecomment-752960642 accordingly. Now, we have the good number of records. Can you explain the discrepancy regarding the differences when counting the number of image files on the server versus the number of E37 instances?

**Comment by tschleider** [comment]
I realised I did not update this yet. 

Images with more E38 than files on server:
- VAM
- MET
- MFA
- IMATEX
- Venetian

These are all crawling / scraping issues. I discussed this with @ehrhart (especially VAM), but there is nothing to be done anymore for the remaining cases. It's a problem on the museums side, as it is links that are provided (by the API), but just don't work, like:

Object from API: https://www.vam.ac.uk/api/json/museumobject/O11067
Object page: http://collections.vam.ac.uk/item/O11067/length-of-silk-unknown/
Image link: http://media.vam.ac.uk/media/thira/collection_images/2020MP/2020MP4349.jpg  

Images with more files than E38:
- Only Garin

As mentioned above we are dealing with messy data that is provided to us in form of excel sheets (including images) and image files, supposedly connected by common file names and the same as inside the excel sheets. Sometimes the file name of the excel sheet does not match with its internal museum ID. We created a matching system in the converter to connect image files with these excel sheets, but we uploaded all image files, even if we could not match them. 

We do not actually extract the images from the excel sheets itself again, as the image file names contain also meta data (ANV, REV or DET e.g.) that we would otherwise lose.  

I simply suspect that we have "too many" image files that do not all fit to the record files.
